### DISCUSSION TITLE
Unregistering routes

### DISCUSSION THREAD

**Discussion by danielepolencic** [honojs/hono#1507] [Ideas]
Is it possible to unregister routes?

Example:

```
app.on('GET', '/temp', handler|middleware...)
```

after some time:

```
app.off('GET', '/temp')
```

The use case is: that we have actors that are spun and destroyed dynamically.

**Comment by revskill10** [comment]
I think you can "hack" by overriding with a `noOpHandler` ?

**Reply by CyanChanges** [reply]
But it's still there and what if you want to add a same or similar route?

**Comment by maou-shonen** [comment]
Why not use middleware?

**Reply by CyanChanges** [reply]
Isn't that same, you can't unregister it and it must have side-effects: the Memory leak

**Reply by maou-shonen** [reply]
Although I don't think there's much difference, I'll still try to answer your question. This should be able to be handled by proxying to another app, like this.
```ts
import { Hono } from "hono"

let subApp = new Hono().get("/tmp", (c) => c.text("tmp"))

const app = new Hono()
  .get("/", (c) => c.text("Hello, World!"))
  // at last
  .all("/*", (c) => {
    return subApp.request(c.req.url, {
      method: c.req.method,
    })
  })

app.request("/tmp") // => "tmp"

subApp = new Hono()

app.request("/tmp") // => 404
```

**Reply by CyanChanges** [reply]
> Although I don't think there's much difference, I'll still try to answer your question. This should be able to be handled by proxying to another app, like this.
> 
> ```ts
> import { Hono } from "hono"
> 
> let subApp = new Hono().get("/tmp", (c) => c.text("tmp"))
> 
> const app = new Hono()
>   .get("/", (c) => c.text("Hello, World!"))
>   // at last
>   .all("/*", (c) => {
>     return subApp.request(c.req.url, {
>       method: c.req.method,
>     })
>   })
> 
> app.request("/tmp") // => "tmp"
> 
> subApp = new Hono()
> 
> app.request("/tmp") // => 404
> ```

But you need to create a new Hono App
So you have to save the routes you need
And create a new Hono App and register it all back
Which took time and lowered the performance

**Reply by maou-shonen** [reply]
This can be completely customized based on your needs... for example, storing an App for each route or something like that...

**Reply by CyanChanges** [reply]
> This can be completely customized based on your needs... for example, storing an App for each route or something like that...

So storing app for each cause memory leak(because you can't delete rout in root app), storing routes and recreate when change decrease the performance?

**Reply by CyanChanges** [reply]
Damn, looks like i am implementing a libc `mallloc()` my self

**Reply by maou-shonen** [reply]
This won't cause a memory leak. I don't understand why you're limiting the solution to the router.

**Reply by CyanChanges** [reply]
> This won't cause a memory leak. I don't understand why you're limiting the solution to the router.

For example, a plugin(actor) registered on "/web"
And it created a new Hono app and register it on the root app
And then the plugin get removed, and what to do? replace the original app with a new do nothing app? 
The do nothing app costs memory. And it would live forever. Because root Hono app have referred it.
What if the plugin load again, what if the plugin have a sub-plugin register on "/web/api"? 
There will be a bunch of do nothing hono app costing memory.

**Comment by yusukebe** [comment]
How about using `except` in the Combine Middleware?
https://hono.dev/docs/middleware/builtin/combine#except
